### PR TITLE
[df] Improve user interface in DistrRDF 

### PR DESCRIPTION
**Pull request by martamaja10**
Add a number of functions that could replace the way user interacts with the C++ elements of the distributed, fully pythonic RDF analysis, so that this is more natural and clear as well as less error prone compared to the previous method. 
The new functions are: 
- DeclareHeaders
- DeclareSharedLibs 
- DeclareFiles 
- DeclareCppCode

The functions work with both Spark and Dask backends. 

The tests are introduced in roottest PR: https://github.com/root-project/roottest/pull/1177

### CODE REVIEW

File: bindings/experimental/distrdf/python/DistRDF/Backends/Base.py [hunk 1]
```diff
@@ -193,9 +198,85 @@ def register_initialization(cls, fun, *args, **kwargs):
...
+        fun(*args, **kwargs) 
+     
+    # TODO and check - make sure this works correctly for the multiple separate declares   
+    @classmethod
+    def register_declaration(cls, declaration): 
+        
+        declaration_string = cls.declaration_str + declaration
+        
+        cls.declaration_str += declaration_string
+        code_to_declare_with_guards = "#ifndef CODE\n#define CODE\n{}\n#endif\n".format(declaration_string)
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Base.py:211]: Here we can just leave the `cls.declaration_str += declaration_string` and use the class attribute in the next line without the extra variable

File: bindings/experimental/distrdf/python/DistRDF/Backends/Base.py [hunk 2]
```diff
@@ -193,9 +198,85 @@ def register_initialization(cls, fun, *args, **kwargs):
...
+
+        else:
+            for path_string in paths_to_shared_libraries:
+                pcm, libraries = Utils.check_pcm_in_library_path(
+                    path_string
+                ) 
+                libraries_to_distribute.update(libraries)
+                pcms_to_distribute.update(pcm)  
+            
+        Utils.declare_shared_libraries(libraries_to_distribute)
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Base.py:240]: Here we should probably rename to `distribute_shared_libraries`

File: bindings/experimental/distrdf/python/DistRDF/Backends/Base.py [hunk 3]
```diff
@@ -193,9 +198,85 @@ def register_initialization(cls, fun, *args, **kwargs):
...
+        headers_to_distribute = set()
+        
+        if isinstance(paths_to_headers, str):
+            headers_to_distribute = (Utils.get_paths_set_from_string(paths_to_headers))
+        else: 
+            for path_to_header in paths_to_headers:
+                sanatized_path_to_header = Utils.get_paths_set_from_string(path_to_header)
+                headers_to_distribute.update(sanatized_path_to_header)
+        
+        Utils.declare_headers(headers_to_distribute)
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Base.py:256]: Also probably should be renamed to `distribute_headers`

File: bindings/experimental/distrdf/python/DistRDF/Backends/Dask/Backend.py [hunk 4]
```diff
@@ -140,8 +142,10 @@ def dask_mapper(current_range: Tuple,
         """
         # Retrieve the current worker local directory
         localdir = get_worker().local_directory
-
-        # Get and declare headers on each worker
+        # print("Local dir: ")
+        # print(localdir)
+
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Dask/Backend.py:147]: Leftover debug statements

File: bindings/experimental/distrdf/python/DistRDF/__init__.py [hunk 5]
```diff
@@ -47,6 +49,113 @@ def initialize(fun, *args, **kwargs):
...
+            files_to_distribute = (Utils.get_paths_set_from_string(paths_to_files))
+        else: 
+            for path_to_file in paths_to_files:
+                sanatized_path_to_file = Utils.get_paths_set_from_string(path_to_file)
+                files_to_distribute.update(sanatized_path_to_file)
+
+        df._headnode.backend.distribute_unique_paths(files_to_distribute)    
+
+    
+def DistributeSharedLib(paths_to_shared_libraries: Iterable[str], df=None) -> None:
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/__init__.py:121]: Also for coherency it should be
```suggestion
def DistributeSharedLibs(paths_to_shared_libraries: Iterable[str], df=None) -> None:
```

**Review by vepadulano — CHANGES_REQUESTED** [review]
Very nice thanks! I left a few comments inline, then I have the following generic remarks:

* An important part of this PR is deciding the naming scheme for the new user-facing API. I see two main ways: being pedantic or being coherent. Pedantic would mean that the function for headers should be `IncludeHeaders`, the one for shared libraries probably should be `LoadSharedLibs` etc. Instead if we want to be coherent and also hint at the fact that this is a tool for distributed execution, we could decide to name everything `Distribute*` so that it is already clear in the name of the function that the code/header/file will be uploaded to the workers somehow.
* I believe we need a section in the docs in `RDataFrame.cxx` describing these functions, with examples of usage. Ideally what I would like to see is something like a transition help guide, just to give an example
```python
# If you do this in your local RDataFrame script
ROOT.gInterpreter.Declare("my_code")
df.Define(...)

# Do this in distributed RDF
ROOT.RDF.Distributed.DistributeCode("my_code")
df.Define(...)
```

File: bindings/experimental/distrdf/python/DistRDF/__init__.py [hunk 6]
```diff
@@ -47,6 +49,113 @@ def initialize(fun, *args, **kwargs):
...
+        if isinstance(paths_to_headers, str):
+            headers_to_distribute = (Utils.get_paths_set_from_string(paths_to_headers))
+        else: 
+            for path_to_header in paths_to_headers:
+                sanatized_path_to_header = Utils.get_paths_set_from_string(path_to_header)
+                headers_to_distribute.update(sanatized_path_to_header)
+        
+        Utils.declare_headers(headers_to_distribute)
+        df._headnode.backend.distribute_unique_paths(headers_to_distribute)
+        Base.BaseBackend.headers.update(headers_to_distribute)
```
[the remark this answers is not in the document]
Comment (martamaja10) [reply bindings/experimental/distrdf/python/DistRDF/__init__.py:90]: That's true, I created a helper function in Utils now (similarly for the shared libraries and files)

**Comment by martamaja10** [comment]
> * An important part of this PR is deciding the naming scheme for the new user-facing API. I see two main ways: being pedantic or being coherent. Pedantic would mean that the function for headers should be `IncludeHeaders`, the one for shared libraries probably should be `LoadSharedLibs` etc. Instead if we want to be coherent and also hint at the fact that this is a tool for distributed execution, we could decide to name everything `Distribute*` so that it is already clear in the name of the function that the code/header/file will be uploaded to the workers somehow.

Personally, I would go for the `Distribute*` approach but we can discuss this next week in person. 

> * I believe we need a section in the docs in `RDataFrame.cxx` describing these functions, with examples of usage. Ideally what I would like to see is something like a transition help guide, just to give an example
> 
> ```python
> # If you do this in your local RDataFrame script
> ROOT.gInterpreter.Declare("my_code")
> df.Define(...)
> 
> # Do this in distributed RDF
> ROOT.RDF.Distributed.DistributeCode("my_code")
> df.Define(...)
> ```

sure, I will do that next. 

I implemented all your comments (but the leftover debug statements - on purpose, as there are a few things we still need to debug - see my comments in the roottest PR - especially multiple declarations).

File: bindings/experimental/distrdf/python/DistRDF/__init__.py [hunk 7]
```diff
@@ -47,6 +49,53 @@ def initialize(fun, *args, **kwargs):
     from DistRDF.Backends import Base
     Base.BaseBackend.register_initialization(fun, *args, **kwargs)
 
+def DeclareCppCode(code_to_declare: str) -> None:
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/__init__.py:52]: I would be lenient towards naming this `DistributeCppCode` just to be consistent with the other new functions

File: tree/dataframe/src/RDataFrame.cxx [hunk 8]
```diff
@@ -932,6 +931,63 @@ LiveVisualize(plot_callback_dict, write_to_tfile)
...
+
+The cpp code is always available to all dataframes. 
+
+~~~{.py}
+# Local RDataFrame script
+ROOT.gInterpreter.Declare("my_code")
+df.Define(...)
+
+# Distributed RDF script
+ROOT.RDF.Experimental.Distributed.DistributeCppCode("my_code")
```
Comment (vepadulano) [inline tree/dataframe/src/RDataFrame.cxx:977]: Indeed the function uses the `Distribute` naming also here in the docs

File: bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py [hunk 9]
```diff
@@ -118,16 +130,31 @@ def spark_mapper(current_range):
...
+            
+            files_on_executor = [
+                pyspark.SparkFiles.get(ntpath.basename(filepath))
+                for filepath in files
+            ]
+                        
+            pcms_on_ex = [
+                pyspark.SparkFiles.get(ntpath.basename(filepath))
+                for filepath in pcms
+            ]
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py:150]: Do we actually need to get these file paths here or can we just remove this code?

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py [hunk 10]
```diff
@@ -134,25 +136,93 @@ def check_pcm_in_library_path(shared_library_path: str) -> Tuple[Set[str], Set[s
         list, list: Two lists, the first with all paths to pcm files, the
             second with all paths to shared libraries.
     """
+    
+    # issue when we have, for example ROOT.RDF.Experimental.Distributed.DistributeSharedLibs("folder/helpers_h.so") - this is False and should be True 
+    # therefore this will not work either and instead we check for the file endings of possible shared libraries 
+    # if os.path.isfile(shared_library_path):
+    #     shared_library_dir = os.path.dirname(os.path.abspath(shared_library_path))
+    #     #shared_library_dir = os.path.abspath(shared_library_path)
+    #     #+ "/"
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py:145]: This comment is not completely clear to me, but it does not affect the rest of the function IIUC. Should we keep it, maybe rephrased?

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py [hunk 11]
```diff
@@ -134,25 +136,93 @@ def check_pcm_in_library_path(shared_library_path: str) -> Tuple[Set[str], Set[s
...
+    
+    shared_library_formats = ['.so', '.dll', '.dylib']
+    
+    if shared_library_path.endswith(tuple(shared_library_formats)):
```
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py:149]:
```suggestion
    shared_library_formats = ('.so', '.dll', '.dylib')
    
    if shared_library_path.endswith(shared_library_formats): 
```
This should remove the need for the extra conversion

File: tree/dataframe/src/RDataFrame.cxx [hunk 12]
```diff
@@ -932,6 +931,63 @@ LiveVisualize(plot_callback_dict, write_to_tfile)
 
 \warning The Live Visualization feature is only supported for the Dask backend.
 
+### Injecting C++ code and using external files into distributed RDF script
+
+Distributed RDF provides an interface for the users who want to inject the C++ code (via header files, shared libraries or declare the code directly)
+into their distributed RDF application, or their application needs to use information from external files which should be distributed 
+to the workers (for example, a JSON or a txt file with necessary parameters information). 
+
+The examples below show the usage of these interface functions: firstly, how this is done in a local PyROOT based
```
Comment (vepadulano) [inline tree/dataframe/src/RDataFrame.cxx:940]:
```suggestion
The examples below show the usage of these interface functions: firstly, how this is done in a local Python
```

**Review by vepadulano — CHANGES_REQUESTED** [review]
Nice, we're getting very close now! Some more minor comments from my side

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py (hunk 10 above)
Comment (martamaja10) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py:145]: I think this comment can be removed - it was more an explanation for me for why this had to be changed to the way it is done now vs how it was done before some tests.

File: bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py (hunk 9 above)
Comment (martamaja10) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py:150]: let me double check

Comment (martamaja10) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py:150]: ok, so for the files_on_executor - we can remove this, not needed 
but for the pcms the issue happens when we want to load a shared library if it comes from the CompileMacro, then we get a lot of warnings of the type: 
```
Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libEG_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libFitPanel_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libFoam_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGCocoa_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGed_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGenVector_G__GenVector32_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGenVector_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGenetic_rdict.pcm

Info in <TCling::LoadPCM>: In-memory ROOT PCM candidate /Users/martaczurylo/Work/distributed_impr/build/lib/libGeomPainter_rdict.pcm
```
and we suppress them by adding the pcms_on_ex. 

I should probably add a test for such a scenario of distributing a shared library too.

**Review by vepadulano — APPROVED** [review]
This is a substantial improvement to the interface of distributed RDataFrame, congrats!
